### PR TITLE
Vendor WebGPU Bindings from web_sys

### LINKED ISSUE #5258: Document web-sys version/API mismatch issues
**Description**
Building a wgpu (version 0.19.1) project with wasm-bindgen results in an error: 
  ```
   Compiling wgpu v0.19.1
  error[E0061]: this function takes 1 argument but 3 arguments were supplied
     --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.19.1/src/backend/webgpu.rs:375:22
      |
  375 |     let mut mapped = web_sys::GpuDepthStencilState::new(
      |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  376 |         map_compare_function(desc.depth_compare),
      |         ---------------------------------------- unexpected argument of type `GpuCompareFunction`
  377 |         desc.depth_write_enabled,
      |         ------------------------ unexpected argument of type `bool`
      |
  note: associated function defined here
     --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/web-sys-0.3.68/src/features/gen_GpuDepthStencilState.rs:27:12
      |
  27  |     pub fn new(format: GpuTextureFormat) -> Self {
      |            ^^^
  help: remove the extra arguments
      |
  376 -         map_compare_function(desc.depth_compare),
  376 +         map_texture_format(desc.format),
      |
  
  error[E0061]: this function takes 1 argument but 2 arguments were supplied
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.19.1/src/backend/webgpu.rs:1693:13
       |
  1693 |             web_sys::GpuVertexState::new(desc.vertex.entry_point, &module.0);
       |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ -------------------------
       |                                          |
       |                                          unexpected argument of type `&str`
       |                                          help: remove the extra argument
       |
  note: associated function defined here
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/web-sys-0.3.68/src/features/gen_GpuVertexState.rs:27:12
       |
  27   |     pub fn new(module: &GpuShaderModule) -> Self {
       |            ^^^
  
  error[E0061]: this function takes 2 arguments but 3 arguments were supplied
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.19.1/src/backend/webgpu.rs:1768:17
       |
  1768 |                 web_sys::GpuFragmentState::new(frag.entry_point, &module.0, &targets);
       |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ----------------             -------- unexpected argument of type `&Array`
       |                                                |
       |                                                expected `&GpuShaderModule`, found `&str`
       |
       = note: expected reference `&GpuShaderModule`
                  found reference `&str`
  note: associated function defined here
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/web-sys-0.3.68/src/features/gen_GpuFragmentState.rs:27:12
       |
  27   |     pub fn new(module: &GpuShaderModule, targets: &::wasm_bindgen::JsValue) -> Self {
       |            ^^^
  help: remove the extra argument
       |
  1768 -                 web_sys::GpuFragmentState::new(frag.entry_point, &module.0, &targets);
  1768 +                 web_sys::GpuFragmentState::new(/* &GpuShaderModule */, &module.0);
       |
  
  error[E0061]: this function takes 1 argument but 2 arguments were supplied
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.19.1/src/backend/webgpu.rs:1793:13
       |
  1793 |             web_sys::GpuProgrammableStage::new(desc.entry_point, &shader_module.0);
       |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ------------------
       |                                                |
       |                                                unexpected argument of type `&str`
       |                                                help: remove the extra argument
       |
  note: associated function defined here
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/web-sys-0.3.68/src/features/gen_GpuProgrammableStage.rs:27:12
       |
  27   |     pub fn new(module: &GpuShaderModule) -> Self {
       |            ^^^
  
  error[E0599]: no method named `write_timestamp` found for struct `GpuCommandEncoder` in the current scope
      --> /home/simon/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.19.1/src/backend/webgpu.rs:2505:14
       |
  2503 | /         encoder_data
  2504 | |             .0
  2505 | |             .write_timestamp(&query_set_data.0, query_index);
       | |             -^^^^^^^^^^^^^^^ method not found in `GpuCommandEncoder`
       | |_____________|
       | 
  
  Some errors have detailed explanations: E0061, E0599.
  For more information about an error, try `rustc --explain E0061`.
  error: could not compile `wgpu` (lib) due to 5 previous errors
  ```

I think this is due to a change in the `web-sys` API.
[Version  0.3.67](https://docs.rs/web-sys/0.3.67/web_sys/struct.GpuDepthStencilState.html) has the expected API:
```
pub fn [new](https://docs.rs/web-sys/0.3.67/web_sys/struct.GpuDepthStencilState.html#method.new)(
    depth_compare: [GpuCompareFunction](https://docs.rs/web-sys/0.3.67/web_sys/enum.GpuCompareFunction.html),
    depth_write_enabled: [bool](https://doc.rust-lang.org/nightly/std/primitive.bool.html),
    format: [GpuTextureFormat](https://docs.rs/web-sys/0.3.67/web_sys/enum.GpuTextureFormat.html)
) -> Self

Construct a new GpuDepthStencilState.

This API requires the following crate features to be activated: GpuCompareFunction, GpuDepthStencilState, GpuTextureFormat

This API is unstable and requires --cfg=web_sys_unstable_apis to be activated, as [described in the wasm-bindgen guide](https://rustwasm.github.io/docs/wasm-bindgen/web-sys/unstable-apis.html)
```
[Version 0.3.68](https://docs.rs/web-sys/0.3.68/web_sys/struct.GpuDepthStencilState.html) does not:
```
pub fn [new](https://docs.rs/web-sys/0.3.68/web_sys/struct.GpuDepthStencilState.html#method.new)(format: [GpuTextureFormat](https://docs.rs/web-sys/0.3.68/web_sys/enum.GpuTextureFormat.html)) -> Self

Construct a new GpuDepthStencilState.

This API requires the following crate features to be activated: GpuDepthStencilState, GpuTextureFormat

This API is unstable and requires --cfg=web_sys_unstable_apis to be activated, as [described in the wasm-bindgen guide](https://rustwasm.github.io/docs/wasm-bindgen/web-sys/unstable-apis.html)
```

[This commit](https://github.com/rustwasm/wasm-bindgen/commit/b1047882dcc3ac984485a518b1899670efcc3aa8) made the change.

**Repro steps**

Cargo.toml:
```
[package]
name = "wgpu_error"
version = "0.1.0"
edition = "2021"

[dependencies]
wgpu = { version = "0.19.1" }
wasm-bindgen = "0.2.84"

[lib]
path = "src/lib.rs"
crate-type = ["cdylib", "rlib"]
```

run
```
export RUSTFLAGS=--cfg=web_sys_unstable_apis 
cargo build \
    --no-default-features \
    --target wasm32-unknown-unknown \
    --lib
```

**Expected vs observed behavior**
The build is successful  

**Platform**
*OS*: Ubuntu 22.04.2 LTS
*toolchain*:
stable-x86_64-unknown-linux-gnu (default)
rustc 1.76.0 (07dca489a 2024-02-04)

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Connections**

Closes https://github.com/gfx-rs/wgpu/issues/5258
Closes https://github.com/gfx-rs/wgpu/pull/5224

**Description**

This PR:
- adds a script to our xtask which automatically vendors the web_sys bindings from a particular checkout of wasm-bindgen.
- converts our webgpu bindings to those bindings.
- vastly simplifes the xtask code

We _do not_ use it public-ally, so this is not a breaking change. I plan to backport this.

cc @daxpedda 

**Testing**

Manual testing of WebGPU, CI, and building manually.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,6 +39,18 @@ Bottom level categories:
 
 ## Unreleased
 
+### Major Changes
+
+#### Vendored WebGPU Bindings from `web_sys` 
+
+**`--cfg=web_sys_unstable_apis` is no longer needed in your `RUSTFLAGS` to compile for WebGPU!!!**
+
+While WebGPU's javascript api is stable in the browsers, the `web_sys` bindings for WebGPU are still improving. As such they are hidden behind the special cfg `--cfg=web_sys_unstable_apis` and are not available by default. Everyone who wanted to use our WebGPU backend needed to enable this cfg in their `RUSTFLAGS`. This was very inconvenient and made it hard to use WebGPU, especially when WebGPU is enabled by default. Additionally, the unstable APIs don't adhere to semver, so there were repeated breakages.
```
Comment (daxpedda) [inline CHANGELOG.md:48]: The JS API is stable? Or what does "stable in the browsers" mean here exactly?

If the JS API is really stable we could potentially stabilize it in `web-sys` as well, but my assumption was that it is indeed not stable.

**Review by daxpedda — APPROVED** [review]
Getting rid of `--cfg=web_sys_unstable_apis` is going to be a big improvment in the user experience!

File: wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// DO NOT EDIT THIS FILE!
```
Comment (Wumpf) [inline wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs:1]: can you generate a .gitattributes file like this https://github.com/rerun-io/rerun/blob/main/crates/re_types_core/.gitattributes with `linguist-generated=true` for each file? Then the files show up as generated in reviews :)

File: wgpu/src/backend/webgpu/ext_bindings.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//! These contain ideomatic Rust extension traits for various parts of the WebGPU
+//! bindings that are missing, need to be improved, or otherwise need to be different
+//! from the generated web_sys bindings.
+
+use crate::backend::webgpu::webgpu_sys;
+use wasm_bindgen::prelude::*;
+
+/// Extension trait for [`web_sys::Navigator`] and [`web_sys::WorkerNavigator`] to
+/// access the `gpu` property.
+pub trait NavigatorGpu {
```
Comment (Wumpf) [inline wgpu/src/backend/webgpu/ext_bindings.rs:12]: convenient 👌

**Review by Wumpf — APPROVED** [review]
wow, you made this look way more straight forward than I thought it could be. Awesome!!

File: CHANGELOG.md (hunk 1 above)
Comment (cwfitzgerald) [reply CHANGELOG.md:48]: To a first order approximation webgpu is stable, yes. It's a bit more complicated in practice, and I don't think web-sys can stabilize for a while yet.
- All programs that work today on a stable browser will continue to work.
- Features that have not shipped on a stable browser, may yet change.
  - `write_timestamp` was removed fairly recently, this was before chrome shipped the timestamp feature.
- the IDL has no stability guarentees - webgpu hasn't been standardized yet. Things like names of interfaces may change, so long as the js "abi" doesn't.

File: wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs (hunk 2 above)
Comment (cwfitzgerald) [reply wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs:1]: Oh cool! How do you even find out about this stuff :)

**Review by grovesNL — APPROVED** [review]
Really cool, thanks for doing this! I wish vendoring wasn't necessary but it seems like the best path forward until it stabilizes.

File: wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs (hunk 2 above)
Comment (ErichDonGubler) [reply wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs:1]: @cwfitzgerald: I imagine @Wumpf read these GH docs: https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github

Comment (ErichDonGubler) [reply wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs:1]: I wonder if this is also an opportunity to mark our vendored code, too, with this? https://github.com/github-linguist/linguist/blob/master/docs/overrides.md#vendored-code

Comment (cwfitzgerald) [reply wgpu/src/backend/webgpu/webgpu_sys/gen_Gpu.rs:1]: @ErichDonGubler what vendored code were you thinking of?

**Pull request merged**